lib/charms/loki_k8s/v0/charm_logging.py

```diff
@@ -0,0 +1,360 @@
+    sanitized_logging_endponts = []
+    if isinstance(logging_endpoints, str):
+        errors.append("invalid return value: expected Iterable[str], got str")
+    else:
+        for endpoint in logging_endpoints:
+            if isinstance(endpoint, str):
+                sanitized_logging_endponts.append(endpoint)
+            else:
+                errors.append(f"invalid endpoint: expected string, got {endpoint!r}")
+
+    if errors:
+        raise InvalidEndpointsError(
+            f"{charm}.{logging_endpoints_getter} should return an iterable of Loki push-api "
+            "(-compatible) endpoints (strings); "
+            f"ERRORS: {errors}"
+        )
+
+    return sanitized_logging_endponts
+
+
+def _get_server_cert(
+    server_cert_getter: _CertGetterType, self: CharmBase, charm: Type[CharmBase]
+) -> Optional[str]:
+    if isinstance(server_cert_getter, property):
+        server_cert = server_cert_getter.__get__(self)
+    else:  # method or callable
+        server_cert = server_cert_getter(self)
+
+    # we're assuming that the ca cert that signed this unit is the same that has signed loki's
+    if server_cert is None:
+        logger.debug(f"{charm.__name__}.{server_cert_getter} returned None.")
+        logger.warning(
+            "Charm logs are being sent over insecure http because a ca cert is "
+            "not provided to the charm_logging module."
+        )
+        return None
+
+    if not isinstance(server_cert, str) and not isinstance(server_cert, Path):
+        raise ValueError(
+            f"{charm}.{server_cert_getter} should return a valid path to a tls cert file (string | Path)); "
+            f"got a {type(server_cert)!r} instead."
+        )
+
+    sc_path = Path(server_cert).absolute()
+    if not sc_path.exists():
+        raise RuntimeError(
+            f"{charm}.{server_cert_getter} returned bad path {server_cert!r}: " f"file not found."
+        )
+
+    return str(sc_path)
+
+
+def _setup_root_logger_initializer(
+    charm: Type[CharmBase],
+    logging_endpoints_getter: _EndpointGetterType,
+    server_cert_getter: Optional[_CertGetterType],
+    service_name: Optional[str] = None,
+):
+    """Patch the charm's initializer and inject a call to set up root logging."""
+    original_init = charm.__init__
+
+    @functools.wraps(original_init)
+    def wrap_init(self: CharmBase, framework: Framework, *args, **kwargs):
+        original_init(self, framework, *args, **kwargs)
+
+        if not is_enabled():
+            logger.debug("Charm logging DISABLED by env: skipping root logger initialization")
+            return
+
+        logging_endpoints = _get_logging_endpoints(logging_endpoints_getter, self, charm)
+
+        if not logging_endpoints:
+            return
+
+        juju_topology = JujuTopology.from_charm(self)
+        labels = {
+            **juju_topology.as_dict(),
+            "service_name": service_name or self.app.name,
+            "charm_type": type(self).__name__,
```

Comment:
Why do we need charm type if we already have charm name from the topology class?
If the class name changes across an upgrade, data will be fragmented.
Happens rarely, but changes in metadata `.name` are even rarer.

Reply:
+1